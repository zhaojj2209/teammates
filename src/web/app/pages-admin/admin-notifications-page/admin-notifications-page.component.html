<h1>Notifications</h1>
<p>This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.</p>
<div id="notifications-timezone">
  <p>All dates are displayed in {{ guessTimezone }} time.</p>
</div>

<div>
  <button id="btn-add-notification" class="btn btn-success" (click)="isNotificationEditFormExpanded = true"
          [disabled]="isNotificationEditFormExpanded || isNotificationLoading">
          <span *ngIf="currentNotificationEditFormMode === NotificationEditFormMode.ADD else editbtn"><i class="fas fa-plus"></i> Add New Notification</span>
          <ng-template #editbtn><i class="fas fa-cog"></i> Edit Existing Notification</ng-template>
  </button>
</div>

<div class="margin-top-30px" *ngIf="isNotificationEditFormExpanded && !isNotificationLoading" @collapseAnim>
  <tm-notification-edit-form [formMode]="currentNotificationEditFormMode"
                             [(model)]="notificationEditFormModel"
                             [guessTimezone]="guessTimezone"
                             (addNewNotificationEvent)="addNewNotificationHandler()"
                             (editExistingNotificationEvent)="editExistingNotificationHandler()"
                             (cancelEditingNotificationEvent)="initNotificationEditFormModel()"></tm-notification-edit-form>
</div>

<tm-loading-retry [shouldShowRetry]="hasNotificationLoadingFailed" [message]="'Failed to load data'"
                  (retryEvent)="retryLoadingAllData()">
  <div class="margin-top-30px">
    <div *tmIsLoading="isNotificationLoading">
      <div class="alert alert-warning margin-top-30px" role="alert" *ngIf="notificationsTableRowModels.length === 0 && !isNotificationLoading">
        You have not created any notifications yet. Use the form above to create a notification.
      </div>

      <tm-notifications-table *ngIf = "notificationsTableRowModels.length > 0"
                              [guessTimezone]="guessTimezone"
                              [notificationsTableRowModels]="notificationsTableRowModels"
                              [notificationsTableRowModelsSortBy]="notificationsTableRowModelsSortBy"
                              [notificationsTableRowModelsSortOrder]="notificationsTableRowModelsSortOrder"
                              (deleteNotificationEvent)="deleteNotificationHandler($event)"
                              (loadNotificationEditFormEvent)="loadNotificationEditFormHandler($event)"
                              (sortNotificationsTableRowModelsEvent)="sortNotificationsTableRowModelsHandler($event)"></tm-notifications-table>
    </div>
  </div>
  <br/>
</tm-loading-retry>
