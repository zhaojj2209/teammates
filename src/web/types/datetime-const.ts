export enum Hours {
    ZERO = 0,
    TWELVE = 12,
    IN_ONE_DAY = 24,
    IN_THREE_DAYS = 72,
    IN_ONE_WEEK = 168,
}

export enum Seconds {
    IN_ONE_MINUTE = 60,
}

export enum Minutes {
    IN_ONE_HOUR = 60,
}

export enum Milliseconds {
    IN_ONE_SECOND = 1000,
    IN_TEN_SECONDS = 10000,
    IN_ONE_MINUTE = 60000,
    IN_TEN_MINUTES = 600000,
    IN_FIFTEEN_MINUTES = 900000,
    IN_ONE_HOUR = 3600000,
    IN_ONE_DAY = 86400000,
    IN_ONE_WEEK = 604800000,
}

export const getDefaultDateFormat = (): DateFormat => {
    return { year: 0, month: 0, day: 0 };
};

export const getLatestTimeFormat = (): TimeFormat => {
    return { hour: 23, minute: 59 };
};

export const getDefaultTimeFormat = (): TimeFormat => {
    return { hour: 0, minute: 0 };
};

export interface DateFormat {
    year: number;
    month: number;
    day: number;
}

export interface TimeFormat {
    hour: number;
    minute: number;
}
