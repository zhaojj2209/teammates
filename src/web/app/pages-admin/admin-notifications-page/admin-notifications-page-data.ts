import { Notification, NotificationStyle, NotificationTargetUser } from '../../../types/api-output';
import { NotificationEditFormModel } from './notification-edit-form/notification-edit-form-model';

/**
 * Structure for example of notification edit model.
 */
export const EXAMPLE_NOTIFICATION_EDIT_MODEL: NotificationEditFormModel = {
  notificationId: 'notification1',
  shown: false,

  startTime: { hour: 0, minute: 0 },
  startDate: { year: 0, month: 0, day: 0 },
  endTime: { hour: 0, minute: 0 },
  endDate: { year: 0, month: 0, day: 0 },

  style: NotificationStyle.SUCCESS,
  targetUser: NotificationTargetUser.INSTRUCTOR,

  title: 'valid title',
  message: 'valid message',

  isSaving: false,
  isDeleting: false,
};
/**
 * Structure for example of notification.
 */
export const EXAMPLE_NOTIFICATION_ONE: Notification = {
  notificationId: 'notification1',
  startTimestamp: 0,
  endTimestamp: 1,
  createdAt: 0,
  updatedAt: 0,
  style: NotificationStyle.SUCCESS,
  targetUser: NotificationTargetUser.INSTRUCTOR,
  title: 'valid title',
  message: 'valid message',
  shown: false,
};
/**
 * Structure for example of notification.
 */
export const EXAMPLE_NOTIFICATION_TWO: Notification = {
  notificationId: 'notification2',
  startTimestamp: 1554960204,
  endTimestamp: 1554977204,
  createdAt: 1554232400,
  updatedAt: 1554232400,
  style: NotificationStyle.DANGER,
  targetUser: NotificationTargetUser.GENERAL,
  title: 'valid title 2',
  message: 'valid message 2',
  shown: false,
};
