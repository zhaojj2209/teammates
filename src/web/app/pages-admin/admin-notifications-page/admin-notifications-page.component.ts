import { Component, OnInit } from '@angular/core';
import moment from 'moment-timezone';
import { finalize } from 'rxjs/operators';
import { NotificationService } from '../../../services/notification.service';
import { SimpleModalService } from '../../../services/simple-modal.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { TableComparatorService } from '../../../services/table-comparator.service';
import { TimezoneService } from '../../../services/timezone.service';
import { MessageOutput, Notification, Notifications } from '../../../types/api-output';
import { NotificationStyle, NotificationTargetUser } from '../../../types/api-request';
import { SortBy, SortOrder } from '../../../types/sort-properties';
import { SimpleModalType } from '../../components/simple-modal/simple-modal-type';
import { collapseAnim } from '../../components/teammates-common/collapse-anim';
import { ErrorMessageOutput } from '../../error-message-output';
import {
  NotificationEditFormMode,
  NotificationEditFormModel,
} from './notification-edit-form/notification-edit-form-model';
import { NotificationsTableRowModel } from './notifications-table/notifications-table-model';

@Component({
  selector: 'tm-admin-notifications-page',
  templateUrl: './admin-notifications-page.component.html',
  styleUrls: ['./admin-notifications-page.component.scss'],
  animations: [collapseAnim],
})
export class AdminNotificationsPageComponent implements OnInit {

  NotificationEditFormMode = NotificationEditFormMode;

  currentNotificationEditFormMode = NotificationEditFormMode.ADD;
  isNotificationLoading = false;
  hasNotificationLoadingFailed = false;

  isNotificationEditFormExpanded = false;

  notificationEditFormModel: NotificationEditFormModel = {
    notificationId: '',
    shown: false,

    startTime: { hour: 0, minute: 0 },
    startDate: { year: 0, month: 0, day: 0 },
    endTime: { hour: 0, minute: 0 },
    endDate: { year: 0, month: 0, day: 0 },

    style: NotificationStyle.SUCCESS,
    targetUser: NotificationTargetUser.GENERAL,

    title: '',
    message: '',

    isSaving: false,
    isDeleting: false,
  };

  notificationsTableRowModels: NotificationsTableRowModel[] = [];
  notificationsTableRowModelsSortBy = SortBy.NOTIFICATION_CREATE_TIME;
  notificationsTableRowModelsSortOrder = SortOrder.DESC;

  guessTimezone = 'UTC';

  constructor(
    private notificationService: NotificationService,
    private simpleModalService: SimpleModalService,
    private statusMessageService: StatusMessageService,
    private tableComparatorService: TableComparatorService,
    private timezoneService: TimezoneService,
  ) {}

  ngOnInit(): void {
    this.initNotificationEditFormModel();
    this.loadNotifications();
    this.guessTimezone = this.timezoneService.guessTimezone();
  }

  /**
   * Closes form and initializes values for the edit form model.
   */
  initNotificationEditFormModel(): void {
    this.isNotificationEditFormExpanded = false;

    const nearFuture: moment.Moment = moment().add(1, 'hours');
    const tomorrow: moment.Moment = moment().add(1, 'days');
    this.notificationEditFormModel = {
      notificationId: '',
      shown: false,

      startTime: {
        minute: nearFuture.hour() === 0 ? 59 : 0, // for 00:00 midnight, we use 23:59
        hour: nearFuture.hour() === 0 ? 23 : nearFuture.hour(),
      },
      startDate: {
        year: nearFuture.year(),
        month: nearFuture.month() + 1, // moment return 0-11 for month
        day: nearFuture.date(),
      },
      endTime: { hour: 23, minute: 59 },
      endDate: {
        year: tomorrow.year(),
        month: tomorrow.month() + 1, // moment return 0-11 for month
        day: tomorrow.date(),
      },

      style: NotificationStyle.SUCCESS,
      targetUser: NotificationTargetUser.GENERAL,

      title: '',
      message: '',

      isSaving: false,
      isDeleting: false,
    };

    this.currentNotificationEditFormMode = NotificationEditFormMode.ADD;
  }

  /**
   * Loads all notifications from backend.
   */
  loadNotifications(): void {
    this.isNotificationLoading = true;
    this.notificationService.getNotifications()
      .pipe(finalize(() => { this.isNotificationLoading = false; }))
      .subscribe(
        (notifications: Notifications) => {
          notifications.notifications.forEach((notification: Notification) => {
            this.notificationsTableRowModels.push({
              isHighlighted: false,
              notification,
            });
          });
          // sort the list using create time, and allocate the index in ascending order
          // note: order is set to be descending here as it will be reversed later
          this.notificationsTableRowModelsSortOrder = SortOrder.ASC;
          this.sortNotificationsTableRowModelsHandler(SortBy.NOTIFICATION_CREATE_TIME);
        },
        (resp: ErrorMessageOutput) => {
          this.hasNotificationLoadingFailed = true;
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      );
  }

  /**
   * Resets error state and retry loading data.
   */
  retryLoadingAllData(): void {
    this.isNotificationLoading = false;
    this.hasNotificationLoadingFailed = false;
    this.loadNotifications();
  }

  /**
   * Loads notification data into the edit form for updating actions.
   * Will do checks to skip or show warnings if necessary.
   */
  loadNotificationEditFormHandler(notification: Notification): void {
    if (this.isNotificationEditFormExpanded) {
      if (this.notificationEditFormModel.notificationId !== notification.notificationId) {
        // Warns user that the existing edits will be cleared
        this.simpleModalService.openConfirmationModal(
          'Discard unsaved edit?',
          SimpleModalType.WARNING,
          'Warning: If you choose to edit another notification, any unsaved changes will be lost.',
        ).result.then(() => {
          this.loadNotificationEditForm(notification);
        });
      }
      // Not to do anything if the notification is already loaded
    } else {
      // Loads notification data into the edit form without doing any check if the form was closed originally
      this.loadNotificationEditForm(notification);
    }
  }

  /**
   * The actual function to load data into edit form.
   */
  loadNotificationEditForm(notification: Notification): void {
    const startTime = moment(notification.startTimestamp);
    const endTime = moment(notification.endTimestamp);
    this.notificationEditFormModel = {
      notificationId: notification.notificationId,
      shown: notification.shown,

      startTime: {
        minute: startTime.hour() === 0 ? 59 : 0, // for 00:00 midnight, we use 23:59
        hour: startTime.hour() === 0 ? 23 : startTime.hour(),
      },
      startDate: {
        year: startTime.year(),
        month: startTime.month() + 1, // moment return 0-11 for month
        day: startTime.date(),
      },
      endTime: {
        minute: endTime.hour() === 0 ? 59 : 0, // for 00:00 midnight, we use 23:59
        hour: endTime.hour() === 0 ? 23 : endTime.hour(),
      },
      endDate: {
        year: endTime.year(),
        month: endTime.month() + 1, // moment return 0-11 for month
        day: endTime.date(),
      },

      style: notification.style,
      targetUser: notification.targetUser,

      title: notification.title,
      message: notification.message,

      isSaving: false,
      isDeleting: false,
    };

    this.currentNotificationEditFormMode = NotificationEditFormMode.EDIT;
    this.isNotificationEditFormExpanded = true;
  }

  /**
   * Adds a new notification.
   */
  addNewNotificationHandler(): void {
    this.notificationEditFormModel.isSaving = true;

    // Timezone is not specified here so it will be guessed from browser's request.
    const startTime = this.timezoneService.resolveLocalDateTime(
      this.notificationEditFormModel.startDate, this.notificationEditFormModel.startTime,
    );
    const endTime = this.timezoneService.resolveLocalDateTime(
      this.notificationEditFormModel.endDate, this.notificationEditFormModel.endTime,
    );

    this.notificationService.createNotification({
      title: this.notificationEditFormModel.title,
      message: this.notificationEditFormModel.message,
      style: this.notificationEditFormModel.style,
      targetUser: this.notificationEditFormModel.targetUser,
      startTimestamp: startTime,
      endTimestamp: endTime,
    })
    .pipe(finalize(() => { this.notificationEditFormModel.isSaving = false; }))
    .subscribe(
      (notification: Notification) => {
        this.notificationsTableRowModels.unshift({
          isHighlighted: true,
          notification,
        });
        this.initNotificationEditFormModel();
        this.statusMessageService.showSuccessToast('Notification created successfully.');
      },
      (resp: ErrorMessageOutput) => {
        this.statusMessageService.showErrorToast(resp.error.message);
      },
    );
  }

  /**
   * Updates an existing notification.
   */
  editExistingNotificationHandler(): void {
    this.notificationEditFormModel.isSaving = true;

    // Timezone is not specified here so it will be guessed from browser's request.
    const startTime = this.timezoneService.resolveLocalDateTime(
      this.notificationEditFormModel.startDate, this.notificationEditFormModel.startTime,
    );
    const endTime = this.timezoneService.resolveLocalDateTime(
      this.notificationEditFormModel.endDate, this.notificationEditFormModel.endTime,
    );

    this.notificationService.updateNotification({
      title: this.notificationEditFormModel.title,
      message: this.notificationEditFormModel.message,
      style: this.notificationEditFormModel.style,
      targetUser: this.notificationEditFormModel.targetUser,
      startTimestamp: startTime,
      endTimestamp: endTime,
    }, this.notificationEditFormModel.notificationId)
    .pipe(finalize(() => { this.notificationEditFormModel.isSaving = false; }))
    .subscribe(
      (notification: Notification) => {
        this.statusMessageService.showSuccessToast('Notification updated successfully.');

        this.notificationsTableRowModels.forEach((rowModel: NotificationsTableRowModel) => {
          if (rowModel.notification.notificationId === notification.notificationId) {
            rowModel.isHighlighted = true;
            rowModel.notification = notification;
          }
        });

        this.initNotificationEditFormModel();
      },
      (resp: ErrorMessageOutput) => {
        this.statusMessageService.showErrorToast(resp.error.message);
      },
    );
  }

  /**
   * Deletes a notification.
   */
  deleteNotificationHandler(notificationId: string): void {
    this.notificationService.deleteNotification(notificationId)
      .subscribe(
        (msg: MessageOutput) => {
          this.statusMessageService.showSuccessToast(msg.message);
          this.notificationsTableRowModels = this.notificationsTableRowModels.filter(
            (notificationsTableRowModel: NotificationsTableRowModel) => {
              return notificationsTableRowModel.notification.notificationId !== notificationId;
            },
          );
        },
        (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      );
  }

  /**
   * Handles sorting event from the table sub-component.
   */
  sortNotificationsTableRowModelsHandler(sortBy: SortBy): void {
    if (this.notificationsTableRowModelsSortBy === sortBy) {
      // if sorting the same column, reverse the order
      this.notificationsTableRowModelsSortOrder =
        this.notificationsTableRowModelsSortOrder === SortOrder.ASC ? SortOrder.DESC : SortOrder.ASC;
    } else {
      // if different column, change the sortBy value and but sort order will not be changed
      this.notificationsTableRowModelsSortBy = sortBy;
    }
    // before sorting, remove highlights from all rows
    this.notificationsTableRowModels.forEach(
      (notificationsTableRowModel: NotificationsTableRowModel) => {
        notificationsTableRowModel.isHighlighted = false;
      },
    );
    this.notificationsTableRowModels.sort(this.getNotificationsTableRowModelsComparator());
  }

  /**
   * Gets a comparator to sort the row models based on sortBy and sortOrder.
   */
  getNotificationsTableRowModelsComparator():
    ((a: NotificationsTableRowModel, b: NotificationsTableRowModel) => number) {
      return (a: NotificationsTableRowModel, b: NotificationsTableRowModel) => {
        let strA: string;
        let strB: string;
        switch (this.notificationsTableRowModelsSortBy) {
          case SortBy.NOTIFICATION_CREATE_TIME:
            strA = String(a.notification.createdAt);
            strB = String(b.notification.createdAt);
            break;
          case SortBy.NOTIFICATION_TITLE:
            strA = a.notification.title;
            strB = b.notification.title;
            break;
          case SortBy.NOTIFICATION_START_TIME:
            strA = String(a.notification.startTimestamp);
            strB = String(b.notification.startTimestamp);
            break;
          case SortBy.NOTIFICATION_END_TIME:
            strA = String(a.notification.endTimestamp);
            strB = String(b.notification.endTimestamp);
            break;
          case SortBy.NOTIFICATION_TARGET_USER:
            strA = String(a.notification.targetUser);
            strB = String(b.notification.targetUser);
            break;
          case SortBy.NOTIFICATION_STYLE:
            strA = String(a.notification.style);
            strB = String(b.notification.style);
            break;
          default:
            strA = '';
            strB = '';
        }

        return this.tableComparatorService.compare(
          this.notificationsTableRowModelsSortBy, this.notificationsTableRowModelsSortOrder, strA, strB);
      };
  }
}
