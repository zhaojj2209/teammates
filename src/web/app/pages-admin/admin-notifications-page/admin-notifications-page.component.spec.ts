import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import moment from 'moment-timezone';
import { of, throwError } from 'rxjs';
import SpyInstance = jest.SpyInstance;
import { NotificationService } from '../../../services/notification.service';
import { SimpleModalService } from '../../../services/simple-modal.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { TimezoneService } from '../../../services/timezone.service';
import { createMockNgbModalRef } from '../../../test-helpers/mock-ngb-modal-ref';
import { Notification } from '../../../types/api-output';
import { SortBy } from '../../../types/sort-properties';
import { SimpleModalType } from '../../components/simple-modal/simple-modal-type';
import {
  EXAMPLE_NOTIFICATION_EDIT_MODEL,
  EXAMPLE_NOTIFICATION_ONE,
  EXAMPLE_NOTIFICATION_TWO,
} from './admin-notifications-page-data';
import { AdminNotificationsPageComponent } from './admin-notifications-page.component';
import { AdminNotificationsPageModule } from './admin-notifications-page.module';
import { NotificationsTableRowModel } from './notifications-table/notifications-table-model';

describe('AdminNotificationsPageComponent', () => {
  let component: AdminNotificationsPageComponent;
  let fixture: ComponentFixture<AdminNotificationsPageComponent>;
  let notificationService: NotificationService;
  let statusMessageService: StatusMessageService;
  let simpleModalService: SimpleModalService;
  let timezoneService: TimezoneService;

  const notificationTableRowModel1: NotificationsTableRowModel = {
    isHighlighted: true,
    notification: EXAMPLE_NOTIFICATION_ONE,
  };

  const notificationTableRowModel2: NotificationsTableRowModel = {
    isHighlighted: false,
    notification: EXAMPLE_NOTIFICATION_TWO,
  };

  /**
   * Verifies the row model is updated as expected.
   */
  function verifyFirstRowModelEqualToExample(expected: Notification): void {
    expect(component.notificationsTableRowModels[0].notification.notificationId).toEqual(expected.notificationId);
    expect(component.notificationsTableRowModels[0].notification.shown).toBeFalsy();
    expect(component.notificationsTableRowModels[0].notification.targetUser).toEqual(expected.targetUser);
    expect(component.notificationsTableRowModels[0].notification.title).toEqual(expected.title);
    expect(component.notificationsTableRowModels[0].notification.message).toEqual(expected.message);
  }

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        AdminNotificationsPageModule,
        HttpClientTestingModule,
        BrowserAnimationsModule,
      ],
      providers: [
        TimezoneService,
      ],
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(AdminNotificationsPageComponent);
    notificationService = TestBed.inject(NotificationService);
    statusMessageService = TestBed.inject(StatusMessageService);
    simpleModalService = TestBed.inject(SimpleModalService);
    timezoneService = TestBed.inject(TimezoneService);
    component = fixture.componentInstance;
    moment.tz.setDefault('SGT');
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default view', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when notifications are loading', () => {
    component.isNotificationLoading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when notifications failed to load', () => {
    component.hasNotificationLoadingFailed = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when notification edit form expanded for adding notification', () => {
    jest.spyOn(timezoneService, 'guessTimezone').mockReturnValue('Asia/Singapore');
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of());
    component.initNotificationEditFormModel();
    component.loadNotifications();
    fixture.detectChanges();

    const btn: any = fixture.debugElement.nativeElement.querySelector('#btn-add-notification');
    expect(btn.disabled).toBeFalsy();
    expect(btn.textContent).toEqual(' Add New Notification');
    btn.click();
    fixture.detectChanges();

    expect(component.isNotificationLoading).toBeFalsy();
    expect(component.isNotificationEditFormExpanded).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should disable edit button when notification edit form expanded for existent notification', () => {
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of({
      notifications: [EXAMPLE_NOTIFICATION_ONE],
    }));
    component.initNotificationEditFormModel();
    component.loadNotifications();
    component.loadNotificationEditForm(component.notificationsTableRowModels[0].notification);
    fixture.detectChanges();

    const btn: any = fixture.debugElement.nativeElement.querySelector('#btn-add-notification');
    expect(btn.disabled).toBeTruthy();
    expect(btn.textContent).toEqual(' Edit Existing Notification');
  });

  it('should load correct notification for a given API output', () => {
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of({
      notifications: [EXAMPLE_NOTIFICATION_ONE],
    }));

    component.loadNotifications();

    expect(component.notificationsTableRowModels.length).toEqual(1);
    verifyFirstRowModelEqualToExample(EXAMPLE_NOTIFICATION_ONE);
  });

  it('should add notification for all fields filled in', () => {
    jest.spyOn(notificationService, 'createNotification').mockReturnValue(of(EXAMPLE_NOTIFICATION_ONE));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
    .mockImplementation((args: string) => {
      expect(args).toEqual('Successfully created');
    });

    component.addNewNotificationHandler();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(component.isNotificationEditFormExpanded).toBeFalsy();
    expect(component.notificationsTableRowModels.length).toEqual(1);
    verifyFirstRowModelEqualToExample(EXAMPLE_NOTIFICATION_ONE);
  });

  it('should display error message when failed to create notification', () => {
    component.notificationEditFormModel = EXAMPLE_NOTIFICATION_EDIT_MODEL;
    jest.spyOn(notificationService, 'createNotification').mockReturnValue(throwError({
      error: {
        message: 'This is the error message.',
      },
    }));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showErrorToast')
      .mockImplementation((args: string) => {
        expect(args).toEqual('This is the error message.');
      });

    component.addNewNotificationHandler();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(component.notificationsTableRowModels.length).toEqual(0);
  });

  it('should display error message when notification failed to load', () => {
    component.hasNotificationLoadingFailed = false;
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(throwError({
      error: {
        message: 'This is the error message.',
      },
    }));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showErrorToast')
      .mockImplementation((args: string) => {
        expect(args).toEqual('This is the error message.');
      });

    component.loadNotifications();

    expect(spy).toBeCalled();
    expect(component.hasNotificationLoadingFailed).toBeTruthy();
  });

  it('should display warning when attempts to edit another notification when form is open', async () => {
    const promise: Promise<void> = Promise.resolve();
    const modalSpy: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));
    jest.spyOn(notificationService, 'getNotifications').mockReturnValue(of({
      notifications: [EXAMPLE_NOTIFICATION_ONE, EXAMPLE_NOTIFICATION_TWO],
    }));
    component.initNotificationEditFormModel();
    component.loadNotifications();
    component.loadNotificationEditForm(component.notificationsTableRowModels[0].notification);
    fixture.detectChanges();
    expect(component.isNotificationEditFormExpanded).toBeTruthy();

    const expectedMessage: string = component.notificationsTableRowModels[0].notification.message;

    component.loadNotificationEditFormHandler(component.notificationsTableRowModels[0].notification);
    component.notificationEditFormModel.message = 'new message';
    expect(component.notificationEditFormModel.message).toEqual('new message');

    // attempts to load edit form for another notification
    component.loadNotificationEditFormHandler(component.notificationsTableRowModels[1].notification);
    await promise;
    expect(component.notificationsTableRowModels[0].notification.message).toEqual(expectedMessage);
    expect(component.notificationEditFormModel.message)
        .toEqual(component.notificationsTableRowModels[1].notification.message);

    expect(modalSpy).toHaveBeenCalledTimes(1);
    expect(modalSpy).toHaveBeenLastCalledWith('Discard unsaved edit?',
        SimpleModalType.WARNING,
        'Warning: If you choose to edit another notification, any unsaved changes will be lost.');
  });

  it('should delete notfication', () => {
    component.notificationsTableRowModels = [notificationTableRowModel1];
    expect(component.notificationsTableRowModels.length).toEqual(1);
    jest.spyOn(notificationService, 'deleteNotification').mockReturnValue(of({
      message: 'Successfully deleted',
    }));
    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
    .mockImplementation((args: string) => {
      expect(args).toEqual('Successfully deleted');
    });

    component.deleteNotificationHandler(EXAMPLE_NOTIFICATION_EDIT_MODEL.notificationId);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(component.notificationsTableRowModels.length).toEqual(0);
  });

  it('should sort notification by start time', () => {
    component.notificationsTableRowModels = [notificationTableRowModel1, notificationTableRowModel2];
    component.notificationsTableRowModelsSortBy = SortBy.NOTIFICATION_CREATE_TIME;
    component.sortNotificationsTableRowModelsHandler(SortBy.NOTIFICATION_START_TIME);
    expect(component.notificationsTableRowModelsSortBy).toEqual(SortBy.NOTIFICATION_START_TIME);
  });
});
