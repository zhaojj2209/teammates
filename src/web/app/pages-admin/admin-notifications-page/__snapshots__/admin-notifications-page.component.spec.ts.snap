// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminNotificationsPageComponent should snap when notification edit form expanded 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  hasNotificationLoadingFailed="false"
  isNotificationEditFormExpanded={[Function Boolean]}
  isNotificationLoading={[Function Boolean]}
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <i
        class="fas fa-plus"
      />
       Add New Notification 
    </button>
  </div><tm-loading-retry>
    <div
      class="margin-top-30px"
    >
      <tm-loading-spinner>
        <div
          class="loading-container"
        >
          <div
            class="row spinner-border text-primary"
            role="status"
          />
          <div
            class="row text-secondary"
          >
            Loading...
          </div>
        </div>
      </tm-loading-spinner>
    </div>
    <br />
  </tm-loading-retry>
</tm-admin-notifications-page>
`;

exports[`AdminNotificationsPageComponent should snap when notifications are loading 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  hasNotificationLoadingFailed="false"
  isNotificationEditFormExpanded="false"
  isNotificationLoading={[Function Boolean]}
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <i
        class="fas fa-plus"
      />
       Add New Notification 
    </button>
  </div><tm-loading-retry>
    <div
      class="margin-top-30px"
    >
      <tm-loading-spinner>
        <div
          class="loading-container"
        >
          <div
            class="row spinner-border text-primary"
            role="status"
          />
          <div
            class="row text-secondary"
          >
            Loading...
          </div>
        </div>
      </tm-loading-spinner>
    </div>
    <br />
  </tm-loading-retry>
</tm-admin-notifications-page>
`;

exports[`AdminNotificationsPageComponent should snap when notifications failed to load 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  hasNotificationLoadingFailed={[Function Boolean]}
  isNotificationEditFormExpanded="false"
  isNotificationLoading={[Function Boolean]}
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <i
        class="fas fa-plus"
      />
       Add New Notification 
    </button>
  </div><tm-loading-retry>
    <div
      class="text-center"
    >
      <p
        class="text-muted m-1"
      >
        Failed to load data
      </p>
      <button
        class="btn btn-primary"
      >
        <i
          aria-hidden="true"
          class="fas fa-redo"
        />
        Â Retry
      </button>
    </div>
  </tm-loading-retry>
</tm-admin-notifications-page>
`;

exports[`AdminNotificationsPageComponent should snap with default view 1`] = `
<tm-admin-notifications-page
  NotificationEditFormMode={[Function Object]}
  hasNotificationLoadingFailed="false"
  isNotificationEditFormExpanded="false"
  isNotificationLoading={[Function Boolean]}
  notificationEditFormModel={[Function Object]}
  notificationService={[Function NotificationService]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezoneService={[Function TimezoneService]}
>
  <h1>
    Notifications
  </h1><p>
    This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.
  </p><div>
    <button
      class="btn btn-success"
      disabled=""
      id="btn-add-notification"
    >
      <i
        class="fas fa-plus"
      />
       Add New Notification 
    </button>
  </div><tm-loading-retry>
    <div
      class="margin-top-30px"
    >
      <tm-loading-spinner>
        <div
          class="loading-container"
        >
          <div
            class="row spinner-border text-primary"
            role="status"
          />
          <div
            class="row text-secondary"
          >
            Loading...
          </div>
        </div>
      </tm-loading-spinner>
    </div>
    <br />
  </tm-loading-retry>
</tm-admin-notifications-page>
`;
